### PR TITLE
interfaces/builtin: add initial docker interface

### PR DESCRIPTION
**Pull request by jdstrand**
This refreshes https://github.com/snapcore/snapd/pull/1619 with trunk, adds documentation, and shuffles around the policy so that we have the 'docker-support' interface that the daemon should 'plugs' and the 'docker' interface for the daemon to 'slots' and clients to 'plugs'. Adds the adds the 'privileged-containers' attribute to docker-support and limits use of the slot to docker from 'docker' or 'canonical' and use of the 'docker-support' interface to docker from 'docker' or 'canonical'.

### CODE REVIEW

**Comment by jdstrand** [comment]
With these updates, if I sideload the snap, I see in /var/log/syslog:
```
Sep 19 22:09:45 snappy-16-edge-amd64 /usr/lib/snapd/snapd[1184]: handlers.go:98: snap "docker" has bad plugs or slots: docker-privileged (daemon-privileged attribute is reserved for the Docker project and Canonical); docker-daemon (docker interface is reserved for the Docker project and Canonical)
```

For testing purposes I need to remove the devName check. Once I do if I sideload the snap then do:

```
$ sudo iptables -L
$ sudo ip6tables -L
$ sudo snap connect docker:docker-cli docker:docker-daemon
$ sudo snap connect docker:firewall-control ubuntu-core:firewall-control
$ sudo snap connect docker:home ubuntu-core:home
$ sudo service snap.docker.dockerd stop
$ sudo service snap.docker.dockerd start
```

Then I can use a trusty container like so:
```
$ sudo docker pull ubuntu:trusty
$ sudo docker run ubuntu:trusty uptime
 21:07:14 up  4:01,  0 users,  load average: 0.00, 0.04, 0.01
```

If I try to launch a privileged container:
```
$ sudo docker run --privileged ubuntu:trusty uptime
docker: Error response from daemon: Container command 'uptime' could not be invoked.
```

If I connect the 'docker-privileged' interface as outlined in https://github.com/snapcore/snapd/pull/1619#issuecomment-246499074:

```
$ sudo snap connect docker:docker-privileged docker:docker-daemon
$ sudo docker run --privileged ubuntu:trusty uptime
 22:34:33 up  4:25,  0 users,  load average: 0.13, 0.05, 0.01
```

And disconnect:
```
$ sudo snap disconnect docker:docker-privileged docker:docker-daemon
$ sudo docker run --privileged ubuntu:trusty uptime
```

As such, it is operating according to the defined aggreements.

**Comment by jdstrand** [comment]
@lool and @tianon - let me know if there is anything that should change. This is the snap.yaml I am using with @tianon's snap:

```
plugs:
  docker-cli:
    interface: docker
  docker-privileged:
    interface: docker
    daemon-privileged: true
slots:
  docker-daemon:
    interface: docker

apps:
  compose:
    command: command-compose.wrapper
    plugs:
    - docker-cli
    - network
    - home
  docker:
    command: command-docker.wrapper
    plugs:
    - docker-cli
    - network
    - home
  dockerd:
    command: command-dockerd.wrapper
    daemon: simple
    plugs:
    - network-bind
    - firewall-control
    - docker-privileged
    slots:
    - docker-daemon
architectures:
- amd64
confinement: strict
description: Docker complements kernel namespacing with a high-level API which operates
  at the process level. It runs unix processes with strong guarantees of isolation
  and repeatability across servers.
name: docker
summary: Docker Linux container runtime
version: 1.11.2-5+jdstrand3
```

**Comment by jdstrand** [comment]
@lool - it would be most convenient if you uploaded a series 16 snap under the canonical publisher so this could more easily be tested. Thanks!

File: interfaces/builtin/docker.go [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+	return nil
+}
+
+func (iface *DockerInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	snapName := slot.Snap.Name()
+	devName := slot.Snap.Developer
+	// The docker slot can only by used with the docker project and
+	// Canonical
+	if snapName != "docker" || (devName != "canonical" && devName != "docker") {
+		return fmt.Errorf("docker interface is reserved for the Docker project and Canonical")
```
Comment (niemeyer) [inline interfaces/builtin/docker.go]: Can we please reword this as:

`"docker interface is reserved for the upstream docker project"`

File: interfaces/builtin/docker.go [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+	// it needs to be bool and it can only by used with the docker project
+	// and Canonical
+	if v, ok := plug.Attrs["daemon-privileged"]; ok {
+		if _, ok = v.(bool); !ok {
+			return fmt.Errorf("docker plug requires bool with 'daemon-privileged'")
+		}
+		snapName := plug.Snap.Name()
+		devName := plug.Snap.Developer
+		if snapName != "docker" || (devName != "canonical" && devName != "docker") {
+			return fmt.Errorf("daemon-privileged attribute is reserved for the Docker project and Canonical")
```
Comment (niemeyer) [inline interfaces/builtin/docker.go]: Can we please reword this as:

`"daemon-privileged attribute is reserved for the upstream docker project"`

File: interfaces/builtin/docker.go [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+		interfaces.SecuritySecComp,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	daemonPrivileged, _ := plug.Attrs["daemon-privileged"].(bool)
```
Comment (niemeyer) [inline interfaces/builtin/docker.go]: I don't understand this setting. This is a plug attribute, which means it's being exercised by applications connecting _to_ the daemon, but the daemon actually runs on the slot end. So connecting one plug from one snap would change the ability of the daemon to run unconfined for every snap. What am I missing?

Comment (tianon) [inline interfaces/builtin/docker.go]: I think the idea is that it's only meaningful for connecting on Docker itself -- ie, you wouldn't grant "privileged" access to a `docker-compose` snap, but would be able to toggle privileged behavior in the Docker daemon itself by connecting and disconnecting the interface on the `docker` snap (so that for most usage of Docker, it can be left disabled, but the users who need it can enable it).

Comment (jdstrand) [inline interfaces/builtin/docker.go]: This is a plug attribute to be used by the daemon not the cli command. Eg, consider:

```
plugs:
  docker-cli:
    interface: docker
  docker-privileged:
    interface: docker
    daemon-privileged: true
slots:
  docker-daemon:
    interface: docker

apps:
  docker:
    plugs:
    - docker-cli
    - network
    - home
  dockerd:
    daemon: simple
    plugs:
    - network-bind
    - firewall-control
    - docker-privileged
    slots:
    - docker-daemon
```

In this manner, dockerd uses 'daemon-privileged'. We've not done this before in interfaces but I believe it is consistent (I also tried to explain it in docs/interfaces.md). 'plugs: [ docker-privileged ]' isn't much different than if we wanted to do 'plugs: [ bluez ]' for dockerd where dockerd is allowed to connect to the bluez snap via the bluez interface except that in this case, it so happens that the docker interface is providing this. It also isn't much different than the 'plugs: [ network-bind ]' that we are already using, except that it is the docker interface that is providing the policy rather than the core snap. I did it this way because I felt it was important to keep the docker policy in one interface rather than pushing the privileged policy out to docker-privileged-support. If you'd prefer that, I can do it, but the user experience will end up equivalent with the snap connect command.

File: interfaces/builtin/docker.go (hunk 2 above)
Comment (jdstrand) [inline interfaces/builtin/docker.go]: Done

File: interfaces/builtin/docker.go (hunk 1 above)
Comment (jdstrand) [inline interfaces/builtin/docker.go]: Done

File: interfaces/builtin/docker.go (hunk 3 above)
Comment (jdstrand) [inline interfaces/builtin/docker.go]: Also, @tianon is correct -- that is precisely the reasoning.

Comment (niemeyer) [inline interfaces/builtin/docker.go]: Sorry, but the design actually makes even less sense to me now.

We have a very simple interface: "docker". It connects applications that want to talk to docker with a docker daemon. The plug side gets what it needs to talk to the daemon. The slot side gets what it needs to be a daemon. So sweet and clean.

Now, on an alternative universe we have a "docker" interface that allows the docker daemon to connect to itself, so that the consumer of this interface on the plug end becomes privileged. The consumer in this case is not the client of docker anymore, but the daemon. Which also happens to be the provider of this interface. The consumer plug connects to the provider slot to change the behavior of the provider. This is super confusing, and completely unrelated to that first clean description above.

I suggest removing this hack altogether so that we can merge the bulk of this interface, and then working on a better solution to the underlying problem that this is trying to solve on a separate PR.

Comment (jdstrand) [inline interfaces/builtin/docker.go]: Done. This removes support for using 'docker run --privileged'.

Comment (jdstrand) [inline interfaces/builtin/docker.go]: I've implemented the changes and shuffled around the policy so that we have the 'docker-support' interface that the daemon should 'plugs' and the 'docker' interface for the daemon to 'slots' and clients to 'plugs'. Adds the adds the 'privileged-containers' attribute to docker-support and limits use of the slot to docker from 'docker' or 'canonical' and use of the 'docker-support' interface to docker from 'docker' or 'canonical'.

**Comment by jdstrand** [comment]
With the recent changes the docker snap should now use:

```
plugs:
  support:
    interface: docker-support
  privileged:
    interface: docker-support
    privileged-containers: true
  docker-cli:
    interface: docker
slots:
  docker-daemon:
    interface: docker

apps:
  compose:
    command: command-compose.wrapper
    plugs:
    - docker-cli
    - network
    - home
  docker:
    command: command-docker.wrapper
    plugs:
    - docker-cli
    - network
    - home
  dockerd:
    command: command-dockerd.wrapper
    daemon: simple
    plugs:
    - network-bind
    - firewall-control
    - support
    - privileged
    slots:
    - docker-daemon
architectures:
- amd64
confinement: strict
description: Docker complements kernel namespacing with a high-level API which operates
  at the process level. It runs unix processes with strong guarantees of isolation
  and repeatability across servers.
name: docker
summary: Docker Linux container runtime
version: 1.11.2-5+jdstrand4
```

With these updates, if I sideload the snap, I see in /var/log/syslog:
```
Sep 20 18:19:07 snappy-16-edge-amd64 snapd[2160]: 2016/09/20 18:19:07.361196 handlers.go:98: snap "docker" has bad plugs or slots: docker-daemon (docker slot interface is reserved for the upstream docker project); privileged, support (docker-support interface is reserved for the upstream docker project)
```

Indicating that both the slot usage of 'docker' and the plugs usage of 'docker-support' are limited.

For testing purposes I need to remove the devName check. Once I do if I sideload the snap then do:
```
# connect plugs needed for docker-daemon to run
$ sudo snap connect docker:support ubuntu-core:docker-support
$ sudo snap connect docker:firewall-control ubuntu-core:firewall-control
$ sudo snap connect docker:home ubuntu-core:home

# make sure the netfilter modules are loaded (temporary)
$ sudo iptables -L ; sudo ip6tables -L

# restart the daemon
$ sudo service snap.docker.dockerd stop
$ sudo service snap.docker.dockerd start

# connect docker command to docker daemon
$ sudo docker info
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
$ sudo snap connect docker:docker-cli docker:docker-daemon
$ sudo docker info
...
$ sudo docker pull ubuntu:trusty
$ sudo docker run ubuntu:trusty uptime
 18:27:15 up  7:48,  0 users,  load average: 0.42, 0.12, 0.03

# start privileged container
$ sudo docker run --privileged ubuntu:trusty uptime
docker: Error response from daemon: Container command 'uptime' could not be invoked.
$ sudo snap connect docker:privileged ubuntu-core:docker-support
$ sudo docker run --privileged ubuntu:trusty uptime
 19:21:49 up  8:42,  0 users,  load average: 0.16, 0.05, 0.01
```

**Comment by niemeyer** [comment]
Merging based on prior reviews on #1619.
